app/services/role_checker.rb

```diff
@@ -0,0 +1,50 @@
+class RoleChecker
+  attr_reader :user, :records
+
+  EDITOR_ROLES = %w(owner collaborator expert scientist moderator)
+  APPROVER_ROLES = %w(owner collaborator)
+  VIEWER_ROLES = %w(owner collaborator expert scientist tester)
+
+  def initialize(user, records)
+    @user = user
+    @records = records
+    @viewer_project_ids = viewer_project_ids
+  end
+
+  def can_edit?
+    ids = user_role_ids(user.roles, EDITOR_ROLES)
+    check_roles(ids, records)
+  end
+
+  def can_approve?
+    ids = user_role_ids(user.roles, APPROVER_ROLES)
+    check_roles(ids, records)
+  end
+
+  def can_view?
+    ids = user_role_ids(user.roles, VIEWER_ROLES)
+    check_roles(ids, records)
+  end
+
+  def viewer_project_ids
```

Comment:
This is still needed here since it's used in all the `WhateverPolicy::Scope.resolve` methods to filter by project id

Reply:
you should memoize the result of this method call vs calling this in the initializer.

Reply:
Was under the impression that having it in the initializer was doing just that, but I forgot to add it to attr_reader so it would return the memoized version. I'll change the method name and keep the instance var/attr_reader the same and that should do it.